pkg/controller/atlascustomrole/atlascustomrole_controller.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/statushandler"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/version"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/atlas"
+)
+
+type AtlasCustomRoleReconciler struct {
+	Client                      client.Client
+	Log                         *zap.SugaredLogger
+	Scheme                      *runtime.Scheme
+	EventRecorder               record.EventRecorder
+	AtlasProvider               atlas.Provider
+	GlobalPredicates            []predicate.Predicate
+	ObjectDeletionProtection    bool
+	SubObjectDeletionProtection bool
+	independentSyncPeriod       time.Duration
+}
+
+func NewAtlasCustomRoleReconciler(
+	mgr manager.Manager,
+	predicates []predicate.Predicate,
+	atlasProvider atlas.Provider,
+	deletionProtection bool,
+	independentSyncPeriod time.Duration,
+	logger *zap.Logger,
+) *AtlasCustomRoleReconciler {
+	return &AtlasCustomRoleReconciler{
+		Client:                   mgr.GetClient(),
+		Log:                      logger.Named("controllers").Named("AtlasCustomRoles").Sugar(),
+		Scheme:                   mgr.GetScheme(),
+		EventRecorder:            mgr.GetEventRecorderFor("AtlasCustomRoles"),
+		AtlasProvider:            atlasProvider,
+		GlobalPredicates:         predicates,
+		ObjectDeletionProtection: deletionProtection,
+		independentSyncPeriod:    independentSyncPeriod,
+	}
+}
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlascustomroles,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlascustomroles/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlascustomroles,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlascustomroles/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",namespace=default,resources=secrets,verbs=create;update;patch;delete
+// +kubebuilder:rbac:groups="",namespace=default,resources=events,verbs=create;patch
+
+func (r *AtlasCustomRoleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	atlasCustomRole := &akov2.AtlasCustomRole{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, atlasCustomRole)
+	objectNotFound := err != nil && apiErrors.IsNotFound(err)
+	failedToRetrieve := err != nil && !objectNotFound
+
+	switch {
+	case failedToRetrieve:
+		return r.fail(req, err), nil
+	case objectNotFound:
+		return r.notFound(req), nil
+	}
```

Comment:
I believe this is new and I do not see it is only applicable to this particular resource type.

If this is the way we want to go now, we should be consistent across all controllers, which means we need to have a task revisiting and retrofitting this instead of simply fail/terminate when resolving the request into a Kubernetes object.

I would, at the very least create a Jira task so we do not forget.

Reply:
I took it from the https://github.com/mongodb/mongodb-atlas-kubernetes/blob/ad40d6f408409362bda7742087bf3681cc5640f7/pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go#L92-L97 that you refactored

Reply:
CLOUDP-284790